### PR TITLE
[WIP] fix(integrations): Update all relevant info when repo name changes

### PR DESCRIPTION
**Pull request by lobsterkatie**
Right now, we update part of the `Repository` record when the repo name changes, but not all of it.

Still to do/figure out:

- how to handle GitHub vs GitHub enterprise vis-à-vis the URL - I did this based on how we do it in `build_repository_config`, but of course GitHub and GitHub enterprise have different versions of those methods. Should I make an `update_repo_data` function which could live on the GitHub `Webhook` class and get overridden by the GitHub enterprise hooks? There's no centralized place to put it for enterprise, though... Not sure the best way to handle this.

- deal with GitLab and/or Azure, if applicable

### CODE REVIEW

File: src/sentry/integrations/github/webhook.py [hunk 1]
```diff
@@ -73,10 +73,15 @@ def __call__(self, event, host=None):
...
+                # slug (in the form 'github_org_name/repo_name').
                 # This is needed to access the API since `external_id` isn't sufficient.
-                if repo.config.get('name') != event['repository']['full_name']:
-                    repo.config['name'] = event['repository']['full_name']
+                repo_full_name = event['repository']['full_name']
+                # update our data if repo name has changed
+                if repo.config.get('name') != repo_full_name:
+                    repo.config['name'] = repo_full_name
+                    repo.name = repo_full_name
+                    repo.url = u'https://github.com/{}'.format(repo_full_name)
```
Comment (lauryndbrown) [inline src/sentry/integrations/github/webhook.py]: @MeredithAnya do you happen to know what the correct way to get the domain name for github enterprise? Would it just be the domain name in this case? https://github.com/getsentry/sentry/blob/master/src/sentry/integrations/github_enterprise/integration.py#L335

Comment (markstory) [reply src/sentry/integrations/github/webhook.py]: Github sends us a `html_url` as part of the the repository object. Could we use that?

Comment (MeredithAnya) [reply src/sentry/integrations/github/webhook.py]: @lobsterkatie  @lauryndbrown  ^ what mark said - we should be able to use `event['repository']['html_url']` which will work for both GH and GHE

File: src/sentry/integrations/bitbucket/webhook.py [hunk 2]
```diff
@@ -70,8 +70,12 @@ def __call__(self, organization, event):
...
             raise Http404()
 
-        if repo.config.get('name') != event['repository']['full_name']:
-            repo.config['name'] = event['repository']['full_name']
+        # update our data if repo name has changed
+        repo_full_name = event['repository']['full_name']
+        if repo.config.get('name') != repo_full_name:
+            repo.config['name'] = repo_full_name
+            repo.name = repo_full_name
+            repo.url = u'https://bitbucket.org/{}'.format(repo_full_name)
```
Comment (lauryndbrown) [inline src/sentry/integrations/bitbucket/webhook.py]: @lobsterkatie it would be good to do something similar to what we're doing with github then. I'm pretty sure bitbucket sends the url as well. :)

File: src/sentry/integrations/github/webhook.py (hunk 1 above)
Comment (lobsterkatie) [reply src/sentry/integrations/github/webhook.py]: LOL - I actually _had_ that originally, and then wanted to match what we did to configure the integration in the first place. Okay, will go back to it.

File: src/sentry/integrations/bitbucket/webhook.py (hunk 2 above)
Comment (lobsterkatie) [reply src/sentry/integrations/bitbucket/webhook.py]: So, BB has `website`, which it describes as "The URL to the repository's website when the code is hosted for a specific website." Not sure how to parse this, and have not had a chance to set up a test BB repo to play with. Do you know if that's the right data/do you have a setup you can test it with easily?

See: https://confluence.atlassian.com/bitbucket/event-payloads-740262817.html#EventPayloads-entity_repository

Comment (markstory) [reply src/sentry/integrations/bitbucket/webhook.py]: You could use `repository.links.html.href`. It looks like it should have the web link to the repository as well. The `website` key might be the user entered site that the repo is *for*. Like sentry.io instead of bitbucket.org/getsentry/sentry

Comment (lobsterkatie) [reply src/sentry/integrations/bitbucket/webhook.py]: Oh, so _that's_ what they meant by that confusing sentence! Thanks for figuring that out, @markstory!

Meanwhile, unfortunately, I don't think `repository.links.html.href` will get us what we're looking for, at least judging by this:
image
(this is what you turn out to get if you click on "Repository property" underneath the table I linked above)

It seems like all of the urls in `links` are API urls, not the public url of the repo (which is what we're going for here). That said - we actually don't have a "BitBucket Enterprise/BitBucket Server/BitBucket On-Prem" integration the way we do with GitHub. Are we likely to? If not, can we leave it the way it is and not worry about it?

**Comment by lobsterkatie** [comment]
Also, stepping back and looking at the problem we're actually trying to solve... is the webhook the place (or at least, the _only_ place) we need to be solving this? All three customers who are running into this problem are running into it when they:

1) look at the repos they have and see one with the old name, and therefore

2) try to add the one with the new name,

3) see it in the list of repos available to add, but then

4) get the `A repository with that name already exists` error.

So it seems to me that a (maybe _the_?) place to be fixing this is when someone tries to add a repo - instead of just refusing because we already have it (sort of), could we instead update it?

(Running into a meeting but I think I see where we'd do that. Will update this comment with that as soon as I have a chance.)

File: src/sentry/integrations/bitbucket/webhook.py (hunk 2 above)
Comment (markstory) [reply src/sentry/integrations/bitbucket/webhook.py]: That is so bizarre that they would put an API URL in `links.html`, but the docs show it 😢

Comment (lobsterkatie) [reply src/sentry/integrations/bitbucket/webhook.py]: I put it back to being `"bitbucket.com/whatever"` for the moment.

File: src/sentry/plugins/providers/integration_repository.py [hunk 3]
```diff
@@ -27,14 +27,43 @@ def __init__(self, id):
...
+                provider=self.id,
+                external_id=config.get('external_id')
+            )
+            if repo is not None:
+                updated = self.update_repository_data(repo, config)
+                if updated:
+                    return Response(serialize(repo, request.user), status=200)
+                return Response({
+                    'errors': {
+                        '__all__': 'A repository with that name already exists'
```
Comment (lobsterkatie) [inline src/sentry/plugins/providers/integration_repository.py]: This (attempting to add a repo which hasn't changed at all) probably won't ever happen, given https://github.com/getsentry/sentry/pull/9288, but can't hurt to handle it.

Also, now that I look at it, might reverse the order of those responses to reflect their relative prevalence.

File: src/sentry/plugins/providers/integration_repository.py [hunk 4]
```diff
@@ -27,14 +27,43 @@ def __init__(self, id):
...
+        # at this point we know it's a new repo
+        try:
             result = self.build_repository_config(
                 organization=organization,
                 data=config,
             )
         except Exception as e:
             return self.handle_api_error(e)
 
+        # TODO do we need all the try-except business here anymore, or are we
```
Comment (lobsterkatie) [inline src/sentry/plugins/providers/integration_repository.py]: We're catching an `IntegrityError`, which I believe `.create()` only throws in the case of duplicate primary keys/violation of unique constraints. (Correct me if I'm wrong about that.) If that's true, I believe those cases should already have been handled above.

File: src/sentry/plugins/providers/integration_repository.py [hunk 5]
```diff
@@ -122,6 +151,13 @@ def get_repository_data(self, organization, config):
         """
         return config
 
+    def update_repository_data(self, repo, new_data):
+        """
+        Update repo in database if necessary. Return True if update happened,
+        False otherwise.
+        """
+        raise NotImplementedError
```
Comment (lobsterkatie) [inline src/sentry/plugins/providers/integration_repository.py]: I'm leaving this up to to individual integrations to implement since each gets its data from slightly different sources (different attribute names in the API response) and since each has its own `repo.config` structure.

File: src/sentry/plugins/providers/integration_repository.py [hunk 6]
```diff
@@ -27,14 +27,43 @@ def __init__(self, id):
...
+                organization_id=organization.id,
+                provider=self.id,
+                external_id=config.get('external_id')
+            ).first()
+            if repo is not None:
+                updated = self.update_repository_data(repo, config)
+                if not updated:
+                    return Response({
+                        'errors': {
+                            '__all__': 'That repository already exists'
```
Comment (lobsterkatie) [inline src/sentry/plugins/providers/integration_repository.py]: This (attempting to add a repo which hasn't changed at all) probably won't ever happen, given https://github.com/getsentry/sentry/pull/9288, but can't hurt to handle it.

File: src/sentry/integrations/bitbucket/repository.py [hunk 7]
```diff
@@ -39,6 +39,33 @@ def get_repository_data(self, organization, config):
...
+        False otherwise.
+        """
+        updated = False
+
+        new_name = data['identifier']
+        if repo.name != new_name:
+            repo.name = new_name
+            updated = True
+
+        new_url = u'https://bitbucket.org/{}'.format(data['name'])
```
Comment (markstory) [inline src/sentry/integrations/bitbucket/repository.py]: I think you'll need `full_name` here. For example one of my bitbucket repos requires my account in the path as well https://bitbucket.org/markstory/sphinx-contrib based on the bitbucket docs, `name` is just the repo name with no account prefix.

Comment (lobsterkatie) [reply src/sentry/integrations/bitbucket/repository.py]: `data['name']` actually is the `full_name` that comes back from the API - see the second-to-last line of `get_repository_data`.

**Comment by lobsterkatie** [comment]
Gentlebumping this. Feedback? Thoughts? Should I go ahead and do the others?

File: src/sentry/plugins/providers/integration_repository.py [hunk 8]
```diff
@@ -27,14 +27,43 @@ def __init__(self, id):
...
+            # get data via API (so we know it's up to date)
             config = self.get_repository_data(organization, request.DATA)
+
+        except Exception as e:
+            return self.handle_api_error(e)
+
+        else:
+            # check to see if we already have this repository, and update any
+            # data if applicable
+            repo = Repository.objects.filter(
```
Comment (lauryndbrown) [inline src/sentry/plugins/providers/integration_repository.py]: A thought here is why use filter when you could use `get`? Was it to use `.first()`? See my comment on `first` below. `get()` would be fine here.

File: src/sentry/plugins/providers/integration_repository.py [hunk 9]
```diff
@@ -27,14 +27,43 @@ def __init__(self, id):
...
+                organization_id=organization.id,
+                provider=self.id,
+                external_id=config.get('external_id')
+            ).first()
+            if repo is not None:
```
Comment (lauryndbrown) [inline src/sentry/plugins/providers/integration_repository.py]: Just style here. In python people tend to write:
```py
if repo:
```
since anything that is `None` will evaluate to false in python.

**Comment by lauryndbrown** [comment]
@lobsterkatie I don't want to take over the pr in any substantial way, but I see the two solutions you outlined as two separate changes, if I were structuring it, I would do it as two prs. Adding to the webhook is going to be relatively easy for me to approve. Whereas creating new functionality in the repo create flow is going to be a bit more complex to review and will take more time. Just thoughts, if you're interested in getting a more immediate fix for the customer.

File: src/sentry/plugins/providers/integration_repository.py (hunk 4 above)
Comment (lauryndbrown) [reply src/sentry/plugins/providers/integration_repository.py]: I agree. I'm not convinced those two should be separate. Could you give me a sense as to you handled it separately as opposed to a `create_or_update`?

File: src/sentry/plugins/providers/integration_repository.py (hunk 9 above)
Comment (lobsterkatie) [reply src/sentry/plugins/providers/integration_repository.py]: Yup, I know... and that specific thing caused a bug in one of our SDK's just last week. :-)

I've been writing it `if repo:` for years, too. But in light of that bug, I'm finally taking my python instructor's advice from years ago and trying to start being explicit and not relying on falsiness (sp?). 

But ¯\_(ツ)_/¯ - I'm happy to put it back if it's the Sentry Way.

File: src/sentry/plugins/providers/integration_repository.py [hunk 10]
```diff
@@ -27,14 +27,43 @@ def __init__(self, id):
...
+            return self.handle_api_error(e)
+
+        else:
+            # check to see if we already have this repository, and update any
+            # data if applicable
+            repo = Repository.objects.filter(
+                organization_id=organization.id,
+                provider=self.id,
+                external_id=config.get('external_id')
+            ).first()
```
[the remark this answers is not in the document]
Comment (lobsterkatie) [reply src/sentry/plugins/providers/integration_repository.py]: Ah, good call. I read that, but clearly not closely enough, as the ordering bit didn't register. Switched to `[0]`.

File: src/sentry/integrations/bitbucket/repository.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (lobsterkatie) [reply src/sentry/integrations/bitbucket/repository.py]: Done.

File: src/sentry/plugins/providers/integration_repository.py (hunk 5 above)
Comment (lobsterkatie) [reply src/sentry/plugins/providers/integration_repository.py]: Indeed, that (implementing it for all the integrations) was my plan, but wanted to get generalized approval of the idea/method before I did it for all of them. Once things look good with the bitbucket one, I can do the rest.

File: src/sentry/plugins/providers/integration_repository.py (hunk 4 above)
Comment (lobsterkatie) [reply src/sentry/plugins/providers/integration_repository.py]: It's partly because the case of creation was already handled separately from the case of it already existing (in which case we had to go undo the side effects of attempting to create the repo, namely deleting the webhook which got created).

Regardless, I think we need to keep it separate for two reasons:

1) We really have three cases, not two:

    a) the repo doesn't exist - create it
    b) the repo does exist but some of the data is wrong - update it
    c) the repo exists and none of the data is different - send back an error

    Case b) is the one I added. Switching to a `create_or_update` would force b) and c) to be the same case, so we'd have to either always send back an error, or never send back an error (the latter being arguably better, but also a change from current behavior).

2) In any case, even if we wanted to do that, unfortunately the "update" part of `create_or_update` isn't as simple as one might like - different information needs to be updated depending on what integration we're talking about (specifically in the `config` column), so I'm not clear how we'd consolidate the behavior into this central location. (That's why I forced each integration to implement its own `update_repository_data` the same way each one has its own `get_repository_data` and `build_repository_config`.)

    The alternative would be to split creation out into the different integrations also... or to call `build_repository_config` _before_ the possible `create_or_update` (so we'd have the correct `config` column value), but then we're exactly back where we started: `build_repository_config` creates a webhook, and we don't want to do that if all we're doing is updating, so we'd have to go and delete it, just like we're trying to get away from.

I do think it would be simpler to just do it all here in `integration_repository.py` and/or with a `create_or_update`, but I just couldn't figure out how to do that given the above.

File: src/sentry/plugins/providers/integration_repository.py (hunk 8 above)
Comment (lobsterkatie) [reply src/sentry/plugins/providers/integration_repository.py]: I had it as a `.get()`, but all my tests crashed. :-(

(That's because it turns out `.get()` doesn't return `None` if no record is found, it raises a `DoesNotExist` error - see https://docs.djangoproject.com/en/2.1/ref/models/querysets/#get . Intuitively I was hoping that a framework written in python would implement `.get()` basically the same way python itself does, but... )

**Comment by lobsterkatie** [comment]
As for splitting the PR in two - happy to do that if you think it would be better. Just let me know.

File: src/sentry/plugins/providers/integration_repository.py (hunk 5 above)
Comment (lauryndbrown) [reply src/sentry/plugins/providers/integration_repository.py]: To me, this is a nice to have feature rather than something I think the other integrations should be forced to have. But I won't push back hard on this one as long as we're avoiding the notimplemented error

File: src/sentry/plugins/providers/integration_repository.py (hunk 9 above)
Comment (lauryndbrown) [reply src/sentry/plugins/providers/integration_repository.py]: You just have the be cognizant of when it can give you a false positive. So, if you think `0` or` ''` or `[]` or `{}` could be things you would want to accept, then it's ok to just go ahead and say what you mean. But in the case here, repo can be either a Repository model object or in the case you are filtering, then it could be an empty QuerySet. First is giving you a couple of things for free here, one being returning `None` if nothing is returned. So in the case you had originally the only two return options are the model object or None. So, it's very safe to simply say `if repo`.

File: src/sentry/plugins/providers/integration_repository.py (hunk 4 above)
Comment (lauryndbrown) [reply src/sentry/plugins/providers/integration_repository.py]: Ok I see. Let's just get the repo. Like you're doing here. I think that'll be the clearer answer. I had thought of `get_or_create` as well. But ultimately I'm not sure we save quite that much and probably lose in terms of readability.

File: src/sentry/plugins/providers/integration_repository.py (hunk 5 above)
Comment (lauryndbrown) [reply src/sentry/plugins/providers/integration_repository.py]: My personal opinion is that this is an edge case and more of a nice to have than something I'd inherently feel was required by the integrations. But I think if you're wanting to implement and write tests for it that's great. :) I would just urge you to turn it into its separate pr.

**Comment by lauryndbrown** [comment]
@lobsterkatie yeah I'd encourage making two prs in this case.

**Comment by lobsterkatie** [comment]
This got abandoned in favor of https://github.com/getsentry/sentry/pull/12290
